### ISSUE TITLE
Error in combineTCR with single chains (Dev Version)

### ISSUE THREAD

**Issue by ElviMagda** [ncborcherding/scRepertoire#96]
Hi Nick,

first of all, thank you for this great package. It helps us all very much :)

I wanted to analyse my data based on just one single chain (TRB) instead of both. I saw you implemented this option in the dev version by using the argument chain = "TRB" in the combineTCR function.
The option for both chains (chain = "both") works perfectly fine, but not for the single chains.

This is my code:
combined <- combineTCR(contig_list, samples = c("Pre", "Post", "Post", "Post", "Post"), ID = c("S1", "S1", "S2", "S3", "S4"), cells ="T-AB", chain = "TRB", removeNA = FALSE, removeMulti = FALSE, filterMulti = FALSE) 

And this is the error I get:
Error in chain %in% c(chain1, chain2) : object 'chain' not found

Am I missing or doing something wrong here? 

Thanks in advance for your help!

Best,
Elvira

> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.4

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] patchwork_1.1.1    tidyr_1.1.3        dplyr_1.0.7        viridis_0.6.1      viridisLite_0.4.0  scales_1.1.1      
 [7] circlize_0.4.13    sctransform_0.3.2  Seurat_4.0.3       scRepertoire_1.3.2 ggplot2_3.3.5      SeuratObject_4.0.2

**Comment by ncborcherding** [comment]
Hey ElviMagda, 

Thanks for reaching out!

Sorry for the confusion - I implemented the change in `combineTCR()` where you can call a specific chain. But after testing it, I felt like the function shouldn't just eliminate half the data - that the combined list will have both chains. Instead, what I did was implement the **chain** parameter into most of the visualization and other functions (including even `combineExpression()`), so everything downstream should now be able to work with calling a single chain.

I am still testing this change, but let me know if you find any issues or have other questions.

Nick

**Comment by ElviMagda** [comment]
Hi Nick,

it makes sense and I think it is a great way to keep the info in the combined list and be able to work with the desired chain downstream the pipeline. 

I will give it a try then :)

Thanks for your fast response.
Best,
Elvira

**Comment by ElviMagda** [comment]
Hi Nick,

sorry for reopening it. 
I have tried the "chain" parameter in some functions (I have also seen them now in the R documentation), but it doesn't seem to work on my hands yet. I have also downloaded again the dev version (1.3.2, right?), just in case, but I couldn't manage to make it work.

Here the code I used and the error I get:

> quantContig(combined, cloneCall="gene+nt", scale = T, chain = "both")
Error in quantContig(combined, cloneCall = "gene+nt", scale = T, chain = "both") : 
  unused argument (chain = "both")

> quantContig(combined, cloneCall="gene+nt", scale = T, chain = "TRA")
Error in quantContig(combined, cloneCall = "gene+nt", scale = T, chain = "TRA") : 
  unused argument (chain = "TRA")

Thank you in advance!!
Best,
Elvira

**Comment by ncborcherding** [comment]
Hey Elvira, 

Have you installed the most up-to-date dev version?

```
devtools::install_github("ncborcherding/scRepertoire@dev")
```

Based on the error, it looks like the chain parameter is not part of the function, which makes me suspect that you don't have the most up-to-date version. 

But let me know if that's not the case and we can troubleshoot. Thanks

Nick

**Comment by ElviMagda** [comment]
Hey,

I did install it before writing my comment. I tried now again, and I get this:

> devtools::install_github("ncborcherding/scRepertoire@dev")
Skipping install of 'scRepertoire' from a github remote, the SHA1 (5b999a48) has not changed since last install.
  Use `force = TRUE` to force installation

So I think I have the most up-to-date version, right?
I got a new Mac Pro with the M1 chip, and I got some problems to install packages from Bioconductor as it is not supported. But I didn't have any problem with your package from Github so far.
I'm very grateful for your help!

Best,
Elvira

**Comment by ncborcherding** [comment]
Hey Elvira, 

Interesting! Sorry for the issue! You should be up-to-date version wise. What does your documentation look like in regards to ```quantContig()```? Below is mine.

<img width="705" alt="Screen Shot 2021-08-08 at 1 41 52 PM" src="https://user-images.githubusercontent.com/22754118/128642385-312f7602-2b5f-4b93-b962-7736cfc85be0.png">

Could you also give me the output of ```sessionInfo()```?

Thanks, 
Nick

**Comment by ElviMagda** [comment]
Hi Nick,

here you go :)

> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.4

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] svglite_2.0.0      patchwork_1.1.1    tidyr_1.1.3        dplyr_1.0.7       
 [5] viridis_0.6.1      viridisLite_0.4.0  scales_1.1.1       circlize_0.4.13   
 [9] sctransform_0.3.2  Seurat_4.0.3       scRepertoire_1.3.2 ggplot2_3.3.5     
[13] SeuratObject_4.0.2

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                  reticulate_1.20             tidyselect_1.1.1           
  [4] htmlwidgets_1.5.3           grid_4.1.0                  Rtsne_0.15                 
  [7] devtools_2.4.2              munsell_0.5.0               codetools_0.2-18           
 [10] ica_1.0-2                   future_1.21.0               miniUI_0.1.1.1             
 [13] withr_2.4.2                 colorspace_2.0-2            Biobase_2.52.0             
 [16] knitr_1.33                  ggalluvial_0.12.3           rstudioapi_0.13            
 [19] stats4_4.1.0                SingleCellExperiment_1.14.1 ROCR_1.0-11                
 [22] tensor_1.5                  listenv_0.8.0               MatrixGenerics_1.4.1       
 [25] labeling_0.4.2              GenomeInfoDbData_1.2.6      polyclip_1.10-0            
 [28] farver_2.1.0                rprojroot_2.0.2             parallelly_1.27.0          
 [31] vctrs_0.3.8                 generics_0.1.0              xfun_0.24                  
 [34] R6_2.5.0                    doParallel_1.0.16           GenomeInfoDb_1.28.1        
 [37] VGAM_1.1-5                  isoband_0.2.5               bitops_1.0-7               
 [40] spatstat.utils_2.2-0        cachem_1.0.5                DelayedArray_0.18.0        
 [43] promises_1.2.0.1            gtable_0.3.0                globals_0.14.0             
 [46] processx_3.5.2              goftest_1.2-2               rlang_0.4.11               
 [49] systemfonts_1.0.2           GlobalOptions_0.1.2         splines_4.1.0              
 [52] lazyeval_0.2.2              spatstat.geom_2.2-2         yaml_2.2.1                 
 [55] reshape2_1.4.4              abind_1.4-5                 httpuv_1.6.1               
 [58] tools_4.1.0                 usethis_2.0.1               cubature_2.0.4.2           
 [61] ellipsis_0.3.2              spatstat.core_2.3-0         RColorBrewer_1.1-2         
 [64] BiocGenerics_0.38.0         sessioninfo_1.1.1           ggridges_0.5.3             
 [67] Rcpp_1.0.7                  plyr_1.8.6                  zlibbioc_1.38.0            
 [70] purrr_0.3.4                 RCurl_1.98-1.3              ps_1.6.0                   
 [73] prettyunits_1.1.1           rpart_4.1-15                deldir_0.2-10              
 [76] pbapply_1.4-3               cowplot_1.1.1               S4Vectors_0.30.0           
 [79] zoo_1.8-9                   SummarizedExperiment_1.22.0 ggrepel_0.9.1              
 [82] cluster_2.1.2               fs_1.5.0                    magrittr_2.0.1             
 [85] data.table_1.14.0           RSpectra_0.16-0             scattermore_0.7            
 [88] SparseM_1.81                lmtest_0.9-38               RANN_2.6.1                 
 [91] truncdist_1.0-2             fitdistrplus_1.1-5          matrixStats_0.60.0         
 [94] pkgload_1.2.1               gsl_2.1-6                   mime_0.11                  
 [97] evaluate_0.14               xtable_1.8-4                IRanges_2.26.0             
[100] gridExtra_2.3               shape_1.4.6                 testthat_3.0.4             
[103] compiler_4.1.0              tibble_3.1.3                KernSmooth_2.23-20         
[106] crayon_1.4.1                htmltools_0.5.1.1           mgcv_1.8-36                
[109] later_1.2.0                 powerTCR_1.12.0             MASS_7.3-54                
[112] Matrix_1.3-4                permute_0.9-5               cli_3.0.1                  
[115] parallel_4.1.0              evd_2.3-3                   igraph_1.2.6               
[118] GenomicRanges_1.44.0        pkgconfig_2.0.3             plotly_4.9.4.1             
[121] spatstat.sparse_2.0-0       foreach_1.5.1               stringdist_0.9.7           
[124] XVector_0.32.0              stringr_1.4.0               callr_3.7.0                
[127] digest_0.6.27               RcppAnnoy_0.0.19            vegan_2.5-7                
[130] spatstat.data_2.1-0         rmarkdown_2.9               leiden_0.3.9               
[133] uwot_0.1.10                 curl_4.3.2                  evmix_2.12                 
[136] shiny_1.6.0                 lifecycle_1.0.0             nlme_3.1-152               
[139] jsonlite_1.7.2              desc_1.3.0                  limma_3.48.1               
[142] fansi_0.5.0                 pillar_1.6.2                lattice_0.20-44            
[145] fastmap_1.1.0               httr_1.4.2                  pkgbuild_1.2.0             
[148] survival_3.2-11             glue_1.4.2                  remotes_2.4.0              
[151] png_0.1-7                   iterators_1.0.13            stringi_1.7.3              
[154] memoise_2.0.0               irlba_2.3.3                 future.apply_1.7.0  
![Screenshot 2021-08-09 at 15 52 55](https://user-images.githubusercontent.com/58399552/128717649-ce18c9f8-b483-4a86-a515-b1b581338e12.png)

**Comment by ncborcherding** [comment]
Hey Elvira, 

Sorry one more question - have you restarted your R session since installing the scRepertoire package? I ask because clearly you have the scRepertoire package installed based on you sessionInfo(), but for some reason the chain parameter is not being recognized. 

This is almost like "chain" parameter doesn't exist, but it should based on your scRepertoire version and your manual page. 
`quantContig(combined, cloneCall="gene+nt", scale = T, chain = "both")
Error in quantContig(combined, cloneCall = "gene+nt", scale = T, chain = "both") :
unused argument (chain = "both")`

I am having trouble recreating the issue - whether I call chain = "TRA" or chain = "both"

<img width="749" alt="Screen Shot 2021-08-09 at 11 35 57 AM" src="https://user-images.githubusercontent.com/22754118/128741655-1f93a0f9-ff1c-42b0-b739-26757b4af425.png">

Sorry this is getting complicated - I am running R 4.05 right now, but I do not see why either that or the new M chip would affect a function parameter? 

Please let me know either way and we can go from there! Thanks,
Nick

**Comment by ElviMagda** [comment]
Hi Nick,

I restarted it and now it works!! 
Thanks a lot for your help and I'm really sorry for any inconvenience. I was working on other pipelines in parallel, that I totally forgot to restart the R session after the installation.
I wish all problems were fixed as easy as this one :)

Thank you again!
Best,
Elvira

**Comment by ncborcherding** [comment]
Great, glad to hear it! 

Let me know if you have any other questions, suggestions or issues!

**Comment by ncborcherding** [comment]
No @dev branch is not in use right now - you should be able to do this using the scRepertoire main/master branch or the current bioconductor release.

Depending on what you really need - you can use `exportClones(..., format = "TCRMatch") or you can use the ```immApex::getIR(..., chains = "TRB") to extract just the B chains from your data. 

Nick
